internal/mongodoc/resource.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package mongodoc // import "gopkg.in/juju/charmstore.v5-unstable/internal/mongodoc"
+
+import (
+	"sort"
+	"time"
+
+	"gopkg.in/errgo.v1"
+	"gopkg.in/juju/charm.v6-unstable"
+	"gopkg.in/juju/charm.v6-unstable/resource"
+	"gopkg.in/juju/charmrepo.v2-unstable/csclient/params"
+	"gopkg.in/mgo.v2/bson"
+)
+
+// NewResourceQuery formats the provided details into a mongo query for
+// the identified resource.
+func NewResourceQuery(cURL *charm.URL, resName string, revision int) bson.D {
+	copied := *cURL
+	cURL = &copied
+	cURL.Revision = -1
+	cURL.Series = ""
+	query := bson.D{
+		{"unresolved-charm-url", cURL},
+		{"name", resName},
+	}
+	if revision >= 0 {
+		query = append(query, bson.DocElem{"revision", revision})
+	}
+	return query
+}
+
+// Resource holds the in-database representation of a charm resource
+// at a particular revision.
+type Resource struct {
+	// CharmURL identifies the unresolved charm associated with this
+	// resource.
+	CharmURL *charm.URL `bson:"unresolved-charm-url"`
+
+	// Name is the name of the resource as defined in the charm
+	// metadata.
+	Name string `bson:"name"`
+
+	// Revision identifies the specific revision of the resource.
+	Revision int `bson:"revision"`
+
+	// Fingerprint is the checksum of the resource file.
+	Fingerprint []byte `bson:"fingerprint"`
+
+	// Size is the size of the resource file, in bytes.
+	Size int64 `bson:"size"`
+
+	// BlobName holds the name that the resource blob is given in the
+	// blob store.
+	BlobName string
+
+	// UploadTime is the UTC timestamp from when the resource file was
+	// stored in the blob store.
+	UploadTime time.Time
+}
+
+// Validate ensures that the doc is valid.
+func (doc Resource) Validate() error {
+	if doc.CharmURL == nil {
+		return errgo.New("missing charm URL")
+	}
+	if doc.CharmURL.Revision != -1 {
+		return errgo.Newf("resolved charm URLs not supported (got revision %d)", doc.CharmURL.Revision)
+	}
+	if doc.CharmURL.Series != "" {
+		return errgo.Newf("series should not be set (got %q)", doc.CharmURL.Series)
+	}
+
+	if doc.Name == "" {
+		return errgo.New("missing name")
+	}
+	if doc.Revision < 0 {
+		return errgo.Newf("got negative revision %d", doc.Revision)
+	}
+
+	if len(doc.Fingerprint) == 0 {
+		return errgo.New("missing fingerprint")
+	}
+	fp, err := resource.NewFingerprint(doc.Fingerprint)
+	if err != nil {
+		return errgo.Notef(err, "bad fingerprint")
+	}
+	if err := fp.Validate(); err != nil {
+		return errgo.Notef(err, "bad fingerprint")
+	}
+
+	if doc.Size < 0 {
+		return errgo.Newf("got negative size %d", doc.Size)
+	}
+
+	if doc.BlobName == "" {
+		return errgo.New("missing blob name")
+	}
+
+	if doc.UploadTime.IsZero() {
+		return errgo.New("missing upload timestamp")
+	}
+
+	return nil
+}
+
+// SortResources sorts the provided resource docs.
+func SortResources(resources []*Resource) {
```

Comment:
While I agree, this is an existing code base with established patterns that I've tried to respect.  Other mongo docs are passed around as pointers.